# FINM 36700 Homework 8

## Andrew Arbitman, Arush Guliani, Tiago Mambrim Flora, Colin Yao

In [73]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression

In [56]:
def load_files(LOADFILE, sheet_name):
  df = pd.read_excel(LOADFILE, sheet_name=sheet_name, index_col = 0)
  return df

## LCTM Case Study

In [57]:
lctm_exhibit_2 = load_files('ltcm_exhibits_data.xlsx', 'Exhibit 2')
lctm_exhibit_4 = load_files('ltcm_exhibits_data.xlsx', 'Exhibit 4')
spy_prices = load_files('spy_data.xlsx', 'prices')
spy_total_returns = load_files('spy_data.xlsx', 'total returns')
spy_excess_returns = load_files('spy_data.xlsx', 'excess returns')

In [58]:
FREQ = 12

In [59]:
#Clean the data
lctm_exhibit_2.columns = [
    "Fund Capital (Billions)", 
    "Gross Monthly Performance", 
    "Net Monthly Performance", 
    "Index of Net Performance"
]
lctm_exhibit_2.drop(lctm_exhibit_2.index[:3], inplace=True)

lctm_exhibit_4.columns = [
    "Total Assets (Billions)",
    "Total Capital (Billions)",
    "Leverage Ratio",
    "Capital Ratio",
]
lctm_exhibit_4.drop(lctm_exhibit_4.index[:2], inplace=True)

### 2. Fund Performance and Attribution

#### 2.1

For both the gross and net series of LTCM excess returns, report the annualized

- mean

- volatility

- Sharpe ratios

Also report the

- skewness

- kurtosis

- 5th quantile

In [63]:
gross_lctm_mean = lctm_exhibit_2["Gross Monthly Performance"].mean() * FREQ
gross_lctm_vol = lctm_exhibit_2["Gross Monthly Performance"].std() * np.sqrt(FREQ)
gross_lctm_sharpe = gross_lctm_mean / gross_lctm_vol
gross_lctm_skewness = lctm_exhibit_2["Gross Monthly Performance"].skew()
gross_lctm_kurtosis = lctm_exhibit_2["Gross Monthly Performance"].kurtosis()
gross_lctm_5th_quantile = lctm_exhibit_2["Gross Monthly Performance"].quantile(0.05)

net_lctm_mean = lctm_exhibit_2["Net Monthly Performance"].mean() * FREQ
net_lctm_vol = lctm_exhibit_2["Net Monthly Performance"].std() * np.sqrt(FREQ)
net_lctm_sharpe = net_lctm_mean / net_lctm_vol
net_lctm_skewness = lctm_exhibit_2["Net Monthly Performance"].skew()
net_lctm_kurtosis = lctm_exhibit_2["Net Monthly Performance"].kurtosis()
net_lctm_5th_quantile = lctm_exhibit_2["Net Monthly Performance"].quantile(0.05)

lctm_stats = pd.DataFrame({
    "Gross Monthly Performance": [gross_lctm_mean, gross_lctm_vol, gross_lctm_sharpe, gross_lctm_skewness, gross_lctm_kurtosis, gross_lctm_5th_quantile],
    "Net Monthly Performance": [net_lctm_mean, net_lctm_vol, net_lctm_sharpe, net_lctm_skewness, net_lctm_kurtosis, net_lctm_5th_quantile]
})
lctm_stats.index = ["Mean", "Volatility", "Sharpe Ratio", "Skewness", "Kurtosis", "5th Quantile"]
lctm_stats

Gross Monthly Performance  Net Monthly Performance
Mean                           0.293887                 0.207170
Volatility                     0.136354                 0.111904
Sharpe Ratio                   2.155321                 1.851315
Skewness                      -0.296428                -0.817870
Kurtosis                       1.569354                 2.905537
5th Quantile                  -0.026400                -0.022400

#### 2.2

Comment on how these stats compare to SPY and other assets we have seen.

How much do they differ between gross and net?

In [65]:
spy_mean = spy_excess_returns["SPY"].mean() * FREQ
spy_vol = spy_excess_returns["SPY"].std() * np.sqrt(FREQ)
spy_sharpe = spy_mean / spy_vol
spy_skewness = spy_excess_returns["SPY"].skew()
spy_kurtosis = spy_excess_returns["SPY"].kurtosis()
spy_5th_quantile = spy_excess_returns["SPY"].quantile(0.05)
spy_stats = pd.DataFrame({
    "SPY": [spy_mean, spy_vol, spy_sharpe, spy_skewness, spy_kurtosis, spy_5th_quantile]
})
spy_stats.index = ["Mean", "Volatility", "Sharpe Ratio", "Skewness", "Kurtosis", "5th Quantile"]
spy_stats

SPY
Mean          0.071563
Volatility    0.149872
Sharpe Ratio  0.477494
Skewness     -0.559686
Kurtosis      0.893472
5th Quantile -0.074058

As you can see, the mean and Sharpe ratio are much better in the LCTM exhibit than for SPY. It also has a better Sharpe ratio than basically every other asset we have seen as a part of this course.

As for the difference between gross and net, it is quite noticable despite it still having great performance. The Sharpe ratio for gross is 2.16 while the Sharpe ratio for net is 1.85. The skewness and kurtosis are very different for net and gross, with gross having a skewness of -0.3 and kurtosis of 1.57 and net having a skewness of -0.82 and a kurtosis of 2.91. 

#### 2.3

Estimate a linear factor decomposition of net LTCM excess returns on SPY excess returns.

Report

- annualized alpha

- beta

- r-squared

Does LTCM deliver performance beyond SPY?

In [ ]:
#Adjust the dates to match, and drop the na's
spy_benchmark = spy_excess_returns.loc["1994-03-01":"1998-07-31"]
lctm_net = lctm_exhibit_2['Net Monthly Performance'].dropna()

In [76]:
y = lctm_net
X = spy_benchmark.values.reshape(-1, 1)
model = LinearRegression()
model.fit(X, y)
alpha = model.intercept_ * FREQ
beta = model.coef_[0]
r_squared = model.score(X, y)
print(f"Annualized Alpha: {alpha:.4f}")
print(f"Beta: {beta:.4f}")
print(f"R-squared: {r_squared:.4f}")

Annualized Alpha: 0.1850
Beta: 0.1430
R-squared: 0.0212


Yes, LCTM delivers performance beyond SPY, with an annualized alpha of 18.5%, which is very large.

#### 2.4

Let’s check for non-linear market exposure. Run the following regression on LTCM’s net excess returns:

Report

- annualized alpha

- the linear and quadratic betas

- r-squared

In [77]:
squared_spy_benchmark = spy_benchmark ** 2
y = lctm_net
X = np.hstack([spy_benchmark, squared_spy_benchmark])
model = LinearRegression()
model.fit(X, y)
alpha = model.intercept_ * FREQ
betas = model.coef_
r_squared = model.score(X, y)
print(f"Annualized Alpha: {alpha:.4f}")
print(f"Linear Beta: {betas[0]:.4f}")
print(f"Quadratic Beta: {betas[1]:.4f}")
print(f"R-squared: {r_squared:.4f}")

Annualized Alpha: 0.2130
Linear Beta: 0.1712
Quadratic Beta: -2.1870
R-squared: 0.0285


#### 2.5

- Does the quadratic market factor do much to increase the overall LTCM variation explained by the market?

- From the regression evidence, does LTCM’s market exposure behave as if it is long market options or short market options?

- Should we describe LTCM as being positively or negatively exposed to market volatility?

The quadratic market factor modestly increases the variation in LTCM returns explained by the market (see the increase in R-squared compared to the linear model), but the improvement is not dramatic. The R^2 of the regression with just the linear market factor is 0.0212 while the R^2 of the regression with both the linear and quadratic market factors is 0.0285.

Based on the negative and significant quadratic beta, LTCM's market exposure behaves as if it is short market options: it benefits less (or loses more) as market returns become more extreme in either direction, which is characteristic of an implicit short volatility or short convexity profile.

Thus, we should describe LTCM as being negatively exposed to market volatility.

#### 2.6

Let’s try to pinpoint the nature of LTCM’s nonlinear exposure. Does it come more from exposure to up-markets or down-markets? Run the following regression on LTCM’s net excess returns:

where k1 = 0.03 and k2 = -0.03.

Report

- annualized alpha

- market beta, the up and down betas

- r-squared

In [85]:
call_option_benchmark = np.maximum((spy_benchmark - 0.03), 0)
put_option_benchmark = np.maximum((-0.03 - spy_benchmark), 0)
y = lctm_net
X = np.hstack([spy_benchmark, call_option_benchmark, put_option_benchmark])
model = LinearRegression()
model.fit(X, y)
alpha = model.intercept_ * FREQ
betas = model.coef_
r_squared = model.score(X, y)
print(f"Annualized Alpha: {alpha:.4f}")
print(f"Market Beta: {betas[0]:.4f}")
print(f"Up Beta: {betas[1]:.4f}")
print(f"Down Beta: {betas[2]:.4f}")
print(f"R-squared: {r_squared:.4f}")

Annualized Alpha: 0.1594
Market Beta: 0.4387
Up Beta: -0.7288
Down Beta: 1.0463
R-squared: 0.0496


#### 2.7

- Is LTCM long or short the call-like factor? And the put-like factor?

- Which factor moves LTCM more, the call-like factor, or the put-like factor?

- In the previous problem, you commented on whether LTCM is positively or negatively exposed to market volatility. Using this current regression, does this volatility exposure come more from being long the market’s upside? Short the market’s downside? Something else?

LCTM’s up beta is negative (-0.73), so it is short the call-like factor, and its down beta is positive (1.05), so it is long the put-like factor.

The absolute value of the down beta (1.05) is larger than that of the up beta (0.73), so LCTM is moved more by the put-like factor.
 
The majority of LCTM’s volatility exposure in this regression comes from being short the market’s upside and long the market’s downside.

## Exercise 8.1

In [209]:
spx_data = load_files('spx_data_weekly.xlsx', 'spx data')
sector_data = load_files('spx_data_weekly.xlsx', 'sector data')
additional_data = load_files('spx_data_weekly.xlsx', 'additional data')
rates_data = load_files('spx_data_weekly.xlsx', 'rates data')

In [210]:
FREQ = 52

In [211]:
#Drop the first two rows of the dataframe
spx_data = spx_data.iloc[2:]

In [213]:
spx_px_last = pd.DataFrame()
spx_eqy_dvd_yld_ind = pd.DataFrame()
spx_pe_ratio = pd.DataFrame()
for i in range(len(spx_data.columns)):
    if i % 3 == 0:
        ticker = spx_data.columns[i]
        spx_px_last[ticker] = spx_data.iloc[:, i]
    elif i % 3 == 1:
        spx_eqy_dvd_yld_ind[ticker] = spx_data.iloc[:, i]
    else:
        spx_pe_ratio[ticker] = spx_data.iloc[:, i]

/var/folders/4h/3bb3wscd33xgrr2xkd9kn6fr0000gn/T/ipykernel_89992/2132115387.py:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  spx_px_last[ticker] = spx_data.iloc[:, i]
/var/folders/4h/3bb3wscd33xgrr2xkd9kn6fr0000gn/T/ipykernel_89992/2132115387.py:9: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  spx_eqy_dvd_yld_ind[ticker] = spx_data.iloc[:, i]
/var/folders/4h/3bb3wscd33xgrr2xkd9kn6fr0000gn/T/ipykernel_89992/2132115387.py:11: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.inser

In [214]:
spx_px_last.head()

A     AAPL     ABBV ABNB      ABT     ACGL  \
2015-07-03 00:00:00  36.4044  28.2589  44.2648  NaN  40.4951  21.6742   
2015-07-10 00:00:00  36.2388  27.5527  44.9267  NaN  40.9213  21.9975   
2015-07-17 00:00:00  36.7447  28.9696   45.757  NaN  41.0942  21.9658   
2015-07-24 00:00:00   36.156  27.8253  44.5083  NaN  42.0413  22.2732   
2015-07-31 00:00:00  37.6645  27.1101  45.7701  NaN  41.7448  22.6187   

                         ACN   ADBE      ADI      ADM  ...       WY     WYNN  \
2015-07-03 00:00:00  81.6872  80.75   52.257   35.923  ...  21.2866  90.2725   
2015-07-10 00:00:00  82.9041  80.59  49.9275  34.8597  ...  20.9189  90.5864   
2015-07-17 00:00:00  84.3309   82.1  50.6986  35.4578  ...  20.3774   88.677   
2015-07-24 00:00:00    84.91  80.98  47.4761  34.8449  ...  19.9027  89.3047   
2015-07-31 00:00:00  86.5382  81.99  47.3463  35.0147  ...  20.5178  90.0022   

                         XEL      XOM      XYL  XYZ      YUM      ZBH    ZBRA  \
2015-07-03 00:00:00  23.8413  52.6941  32.3159  NaN  54.2752  96.5853  112.43   
2015-07-10 00:00:00  24.3716   52.111  31.1369  NaN  53.8662  95.3533  109.75   
2015-07-17 00:00:00  24.4805  52.3582  31.5064  NaN  52.3327  95.9693  114.18   
2015-07-24 00:00:00  23.8703   50.666  30.4682  NaN   51.636  95.2194  110.89   
2015-07-31 00:00:00  25.1852  50.2033  30.3803  NaN  52.2613  92.9072  107.63   

                         ZTS  
2015-07-03 00:00:00  44.5029  
2015-07-10 00:00:00  43.2593  
2015-07-17 00:00:00  44.0331  
2015-07-24 00:00:00    45.59  
2015-07-31 00:00:00  45.1201  

[5 rows x 500 columns]

In [215]:
spx_eqy_dvd_yld_ind.head()

A    AAPL    ABBV ABNB     ABT ACGL     ACN ADBE  \
2015-07-03 00:00:00  1.0988  1.8401  4.6086  NaN  2.3707  NaN  2.4973  NaN   
2015-07-10 00:00:00  1.1038  1.8873  4.5407  NaN   2.346  NaN  2.4607  NaN   
2015-07-17 00:00:00  1.0886   1.795  4.4583  NaN  2.3361  NaN   2.419  NaN   
2015-07-24 00:00:00  1.1063  1.8688  4.5834  NaN  2.2835  NaN  2.4025  NaN   
2015-07-31 00:00:00   1.062  1.9181  4.4571  NaN  2.2997  NaN  2.3573  NaN   

                        ADI     ADM  ...      WY    WYNN     XEL     XOM  \
2015-07-03 00:00:00  3.0618  3.1178  ...  5.4494  2.2155  5.3688  5.5414   
2015-07-10 00:00:00  3.2046  3.2129  ...  5.5452  2.2078   5.252  5.6034   
2015-07-17 00:00:00  3.1559  3.1587  ...  5.6926  2.2554  5.2286   5.577   
2015-07-24 00:00:00  3.3701  3.2142  ...  5.8284  2.2395  5.3623  5.7632   
2015-07-31 00:00:00  3.3794  3.1987  ...  5.6536  2.2222  5.0824  5.8164   

                        XYL  XYZ     YUM     ZBH ZBRA     ZTS  
2015-07-03 00:00:00  1.7428  NaN  3.0216  0.9111  NaN   0.746  
2015-07-10 00:00:00  1.8088  NaN  3.0446  0.9229  NaN  0.7675  
2015-07-17 00:00:00  1.7876  NaN  3.1338   0.917  NaN   0.754  
2015-07-24 00:00:00  1.8485  NaN  3.1761  0.9242  NaN  0.7282  
2015-07-31 00:00:00  1.8538  NaN  3.1381  0.9472  NaN  0.7358  

[5 rows x 500 columns]

In [216]:
spx_pe_ratio.head()

A     AAPL     ABBV ABNB      ABT     ACGL  \
2015-07-03 00:00:00   27.472  13.0592  11.9239  NaN  21.6464  11.8201   
2015-07-10 00:00:00  27.3471  12.7328  12.1022  NaN  21.8742  11.9964   
2015-07-17 00:00:00  27.7288  13.3876  12.3258  NaN  21.9667  11.9791   
2015-07-24 00:00:00  27.2846  12.8588  11.9895  NaN  22.4729  12.1468   
2015-07-31 00:00:00  26.6192  12.5283  12.3294  NaN  22.3145  12.3352   

                         ACN      ADBE      ADI      ADM  ...       WY  \
2015-07-03 00:00:00  17.2055  112.3139  20.1826  10.3845  ...  20.8556   
2015-07-10 00:00:00  17.4618  112.0914  19.2828  10.0771  ...  20.4954   
2015-07-17 00:00:00  17.7623  114.1916  19.5806    10.25  ...  19.9648   
2015-07-24 00:00:00  17.8843  112.6338  18.3361  10.0728  ...  19.4997   
2015-07-31 00:00:00  18.2272  114.0386  17.3337  10.1219  ...  20.1023   

                        WYNN      XEL     XOM      XYL  XYZ      YUM      ZBH  \
2015-07-03 00:00:00  19.6556  12.1268  9.4981  16.9065  NaN  18.6689  18.2975   
2015-07-10 00:00:00   19.724  12.3966   9.393  16.2897  NaN  18.5282  18.0642   
2015-07-17 00:00:00  19.3082   12.452  9.4376  16.4831  NaN  18.0008  18.1809   
2015-07-24 00:00:00  19.4449  12.1416  9.1326  15.9399  NaN  17.7611  18.0388   
2015-07-31 00:00:00  19.5968  12.8104  9.0492  15.8939  NaN  17.9762  17.6008   

                         ZBRA      ZTS  
2015-07-03 00:00:00  151.5166  26.8168  
2015-07-10 00:00:00  147.9049  26.0674  
2015-07-17 00:00:00   153.875  26.5337  
2015-07-24 00:00:00  149.4412  27.4718  
2015-07-31 00:00:00  145.0478  27.1887  

[5 rows x 500 columns]

#### 1.0

Filter the data to drop any ticker which does not have at least 5 years of continuous price data.

In [192]:
def continuous_data_5_years(column) -> bool:
    # Check for a window of 5*FREQ continuous non-NaN values anywhere in the column
    count = 0
    max_count = 0
    for val in column:
        if pd.notna(val):
            count += 1
            if count > max_count:
                max_count = count
        else:
            count = 0
    return max_count >= 5 * FREQ

In [222]:
tickers_to_delete = set()
for ticker in spx_px_last.columns:
    if not continuous_data_5_years(spx_px_last[ticker]):
        tickers_to_delete.add(ticker)
for ticker in spx_eqy_dvd_yld_ind.columns:
    if not continuous_data_5_years(spx_eqy_dvd_yld_ind[ticker]):
        tickers_to_delete.add(ticker)
for ticker in spx_pe_ratio.columns:
    if not continuous_data_5_years(spx_pe_ratio[ticker]):
        tickers_to_delete.add(ticker)
tickers_to_delete = list(tickers_to_delete)

In [223]:
for ticker in tickers_to_delete:
    spx_px_last = spx_px_last.drop(ticker, axis=1)
    spx_eqy_dvd_yld_ind = spx_eqy_dvd_yld_ind.drop(ticker, axis=1)
    spx_pe_ratio = spx_pe_ratio.drop(ticker, axis=1)
print(spx_px_last.shape)
print(spx_eqy_dvd_yld_ind.shape)
print(spx_pe_ratio.shape)

(522, 346)
(522, 346)
(522, 346)


#### 1.1

Report the highest and lowest dividend-yielding stocks…

- for any given date across the entire panel

- taking an average of the past year

Which stocks were they? Was it driven more by changes in D or P?

In [226]:
spx_eqy_dvd_yld_ind.head()

A    AAPL    ABBV     ABT     ACN     ADI     ADM  \
2015-07-03 00:00:00  1.0988  1.8401  4.6086  2.3707  2.4973  3.0618  3.1178   
2015-07-10 00:00:00  1.1038  1.8873  4.5407   2.346  2.4607  3.2046  3.2129   
2015-07-17 00:00:00  1.0886   1.795  4.4583  2.3361   2.419  3.1559  3.1587   
2015-07-24 00:00:00  1.1063  1.8688  4.5834  2.2835  2.4025  3.3701  3.2142   
2015-07-31 00:00:00   1.062  1.9181  4.4571  2.2997  2.3573  3.3794  3.1987   

                        ADP     AEE     AEP  ...     WRB     WSM     WST  \
2015-07-03 00:00:00  3.0251  5.7912  5.6748  ...  1.0943  2.1691  0.7985   
2015-07-10 00:00:00  2.9995  5.7141   5.539  ...  1.0597  2.1402  0.7669   
2015-07-17 00:00:00  2.9679  5.6604  5.5231  ...   1.058  2.1412  0.8212   
2015-07-24 00:00:00   3.065  5.7867  5.6364  ...  1.0576  2.2054  0.8368   
2015-07-31 00:00:00  3.0653  5.4275  5.4401  ...  1.0699  2.1155  0.8358   

                        WTW     XEL     XOM     XYL     YUM     ZBH     ZTS  
2015-07-03 00:00:00  3.0813  5.3688  5.5414  1.7428  3.0216  0.9111   0.746  
2015-07-10 00:00:00  3.0734   5.252  5.6034  1.8088  3.0446  0.9229  0.7675  
2015-07-17 00:00:00  3.0728  5.2286   5.577  1.7876  3.1338   0.917   0.754  
2015-07-24 00:00:00  3.1273  5.3623  5.7632  1.8485  3.1761  0.9242  0.7282  
2015-07-31 00:00:00  3.1065  5.0824  5.8164  1.8538  3.1381  0.9472  0.7358  

[5 rows x 346 columns]

In [243]:
dividend_stats_per_date = []

for date in spx_eqy_dvd_yld_ind.index:
    max_dividend = spx_eqy_dvd_yld_ind.loc[date].max()
    min_dividend = spx_eqy_dvd_yld_ind.loc[date].min()
    max_dividend_stock = spx_eqy_dvd_yld_ind.loc[date].idxmax()
    min_dividend_stock = spx_eqy_dvd_yld_ind.loc[date].idxmin()
    dividend_stats_per_date.append({
        'Date': date,
        'Max Dividend': max_dividend,
        'Min Dividend': min_dividend,
        'Max Dividend Stock': max_dividend_stock,
        'Min Dividend Stock': min_dividend_stock
    })

dividend_stats_per_date = pd.DataFrame(dividend_stats_per_date).set_index('Date')
dividend_stats_per_date

Max Dividend  Min Dividend Max Dividend Stock Min Dividend Stock
Date                                                                        
2015-07-03       20.8790        0.0270                 DD                 CI
2015-07-10       20.8668        0.0277                 DD                 CI
2015-07-17       20.9647        0.0284                 DD                 CI
2015-07-24       23.3362        0.0299                 DD                 CI
2015-07-31       22.7957        0.0302                 DD                 CI
...                  ...           ...                ...                ...
2025-05-30       10.2396        0.0296                DOW               NVDA
2025-06-06        9.9491        0.0282                DOW               NVDA
2025-06-13        9.4999        0.0282                DOW               NVDA
2025-06-20       10.2322        0.0278                DOW               NVDA
2025-06-27       10.5202        0.0254                DOW               NVDA

[522 rows x 4 columns]

In [244]:
dividend_stats_rolling_year = []

for date in spx_eqy_dvd_yld_ind.index:
    if date - spx_eqy_dvd_yld_ind.index[0] >= pd.Timedelta(days=365):
        rolling_year = spx_eqy_dvd_yld_ind[date-pd.Timedelta(days=365):date]
        rolling_means = rolling_year.mean()
        max_dividend = rolling_means.max()
        min_dividend = rolling_means.min()
        max_dividend_stock = rolling_means.idxmax()
        min_dividend_stock = rolling_means.idxmin()
        dividend_stats_rolling_year.append({
            'Date': date,
            'Max Dividend': max_dividend,
            'Min Dividend': min_dividend,
            'Max Dividend Stock': max_dividend_stock,
            'Min Dividend Stock': min_dividend_stock
        })

dividend_stats_rolling_year = pd.DataFrame(dividend_stats_rolling_year).set_index('Date')
dividend_stats_rolling_year

Max Dividend  Min Dividend Max Dividend Stock Min Dividend Stock
Date                                                                        
2016-07-08     23.140868      0.031743                 DD                 CI
2016-07-15     23.157502      0.031845                 DD                 CI
2016-07-22     23.164100      0.031872                 DD                 CI
2016-07-29     23.121832      0.031943                 DD                 CI
2016-08-05     23.089985      0.032013                 DD                 CI
...                  ...           ...                ...                ...
2025-05-30      8.160391      0.032196                 MO               NVDA
2025-06-06      8.115958      0.032040                 MO               NVDA
2025-06-13      8.071968      0.031947                 MO               NVDA
2025-06-20      8.020328      0.031900                 MO               NVDA
2025-06-27      7.976585      0.031783                 MO               NVDA

[469 rows x 4 columns]

This was driven more by changes in D than by changes in P.